### ISSUE TITLE
Unexpected Error: type 'bool' is not a subtype of type 'String' in type cast

### ISSUE THREAD

**Issue by tnedich** [authpass/authpass#178]
**OS:** Win 10 Pro
**Authpass version:** AuthPass 1.7.8 Build 1552

When trying to open file from Google Drive this problem occurs:

`Unexpected Error: type 'bool' is not a subtype of type 'String' in type cast.`

It seems that this error happens all trough the app (For me).

**Comment by hpoul** [comment]
Seems to be related to #154 - can you please see if you have a log file in overflow menu -> About
If not, can you try to open AuthPass.exe through command line, whether you see some log output?

otherwise maybe try the latest beta version: https://data.authpass.app/data/artifacts/AuthPass-setup-1.7.9_1589.exe .. if this also doesn't work you might have to downgrade to https://data.authpass.app/data/artifacts/AuthPass-setup-1.7.7_1519.exe

**Comment by tnedich** [comment]
Unfortunately there is no Log file. When I run it trough cmd no log output. Do I need to run it with some flag.

**Comment by tnedich** [comment]
Same in version 1.7.9 Build 1589

**Comment by drfanatic** [comment]
Same here, for Authpass version 1.7.8 Build 1552 and 1.7.9 Build 1589.

**Comment by fialky** [comment]
Just a confirmation of this error when you try to create or open a key database on Windows 10 x64 with AuthPass-setup-1.7.8_1552:

1) Unexpected Error: type 'bool' is not a subtype of type 'String' in type cast 
2) Error while running task 
    type 'bool' is not a subtype of type 'String' in type cast

**Comment by hpoul** [comment]
It would be awesome if any windows user could try to start this through a development environment to get logs.. I don't have a windows machine at hand.. it should simply work by getting the [latest flutter dev channel](https://flutter.dev/docs/development/tools/sdk/releases?tab=windows), installing [visual studio 2019](https://flutter.dev/desktop#additional-windows-requirements) - [enabling desktop support](https://flutter.dev/desktop#set-up) `flutter config --enable-windows-desktop` .. cloning https://github.com/authpass/authpass.git ([plus submodules](https://github.com/authpass/authpass/blob/master/CONTRIBUTING.md#setup-development-environment) `git submodule update --init`) and doing a `flutter pub get && flutter run -t lib/env/development.dart ` :-)

**Comment by fialky** [comment]
Thank you for your steps, After last command I ended with this text:
.
.
.
flutter: 2020-10-12 23:45:42.061686 FINER analytics - (noop) screen(/onboarding)
flutter: 2020-10-12 23:45:42.211709 WARNING biometric_storage_win32 - Error: 0
### WindowsException: Error 0x00000000: The operation completed successfully.
Syncing files to device Windows...                                 484ms

Flutter run key commands.
r Hot reload.
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on Windows is available at: http://127.0.0.1:61493/3xDyzWqBwng=/
Lost connection to device.

I saw that AuthPass opened for a few seconds and then stopped. New authpass.exe appears in build\windows\runner\Debug. It can be started manually from windows explorer, I can reproduce the error but no log file message in "three dots" menu -> About.

**Comment by hpoul** [comment]
@fialky thanks for giving it a shot.. it's weird that the debug exe can be launched, but it won't start with `flutter run -t lib/env/development.dart` can you try it a second/third time, if anything changes? Have you tried accessing the linked observatory debugger from the output? 🤔️

**Comment by fialky** [comment]
Great, I am able to start it with that command. But in About dialog... _Log file: No Log File?!_ I attach stdout - does it help? I have installed VSCode and something running in it too. If you know where approximately, I can put some breakpoints there :-).
[stdout_log_01.txt](https://github.com/authpass/authpass/files/5368042/stdout_log_01.txt)

**Comment by hpoul** [comment]
@fialky thanks for the detailed debugging and sorry for the late response - i have updated a few dependencies for the desktop embedder, maybe you could give this build a try: https://data.authpass.app/data/artifacts/AuthPass-setup-1.7.9_1601.exe (or update from master, run `flutter pub get` and try again).

**Comment by fialky** [comment]
Nice job! AuthPass-setup-1.7.9_1601 works well on Windows - creating and opening kdbx works both. One minor issue: You have to open it twice. First run - nothing happens (no authpass process in the task manager). Second run - the application appears (authpass process in the task manager presented). If you close it, then you have to open it two times again. (But it could be cuased by flutter itself.)

**Comment by hpoul** [comment]
@fialky thanks for testing.. hmm.. i would suspect that there is still some crash hidden somewhere.. my suspicion is is reading the win32 credential store: https://github.com/authpass/biometric_storage/blob/master/lib/src/biometric_storage_win32.dart#L79-L83 (or better said the cleanup after trying to read, and failing) .. maybe it's wrong to `free` those pointers when reading was not successful.. 🤔️

**Comment by hpoul** [comment]
> First run - nothing happens

@fialky can you reproduce this also in debug mode with `flutter run`? it would be interesting to see the debug output. i'll add even more debugging and try to fix it in the package.. but i can't reproduce it in the package alone..

**Comment by fialky** [comment]
Is it possible to reach you privately by some messenger or by e-mail? I would like to help but without deep knowledge I am clumsy with complex Flutter environment. So, quick share screen session (e.g. MS Teams event) should be very effective and you can get everything what you wish :-). The extraction of relevant steps can be put here later.

**Comment by hpoul** [comment]
i'm usually on the discord channel https://authpass.app/go/discord (email is also okay herbert@authpass.app) .. I usually prefer async communication ;-) but let me know what the problems are and i'll try to help

**Comment by Trekky12** [comment]
Hey @hpoul, I tried to run the current master state with `flutter run` and it is opened after the build process without problems.

There is only a warning regarding the (missing?) biometric storage:
```
flutter: 2020-10-16 16:34:04.178250 WARNING biometric_storage_win32 - Error: 0 
```
I can send you the logfile if you need it.

I have also created a build from the current master and the application is opened without a second try.

**Comment by hpoul** [comment]
@Trekky12 thanks for testing.. that "error" is pretty strange.. because the error code 0 is "success".. https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-

```
      if (CredRead(namePointer, CRED_TYPE_GENERIC, 0, credPointer) != TRUE) {
        final errorCode = GetLastError();
        if (errorCode == ERROR_NOT_FOUND) {
          _logger.fine('Unable to find credential of name $name');
        } else {
          _logger.warning('Error: $errorCode ',
              WindowsException(HRESULT_FROM_WIN32(errorCode)));
        }
```

but I do have the feeling that dart/flutter is doing something strange in `if` statement.. my theory is that the `if` statement itself issues some kind of system call which resets `GetLastError()`.. we might investigate later.. but if it's at least not crashing any more, i'll create a new release.. 1.7.9 is long overdue i guess :-)

**Comment by fialky** [comment]
Hello @hpoul new messages from me finally. I made these updates today and a lot of problems disappeared. (Hence, I postponed a private message.)

1. flutter upgrade
Flutter 1.23.0-19.0.pre.100 • channel master • https://github.com/flutter/flutter.git
Framework • revision 4e2f82cf93 (75 minutes ago) • 2020-10-20 00:19:48 +0900
Engine • revision d03b759d04
Tools • Dart 2.11.0 (build 2.11.0-229.0.dev)

2. authpass clone with submodules
commit 4e91d7cfdb5c365905260bf893d09280b520fb97 (HEAD -> master, origin/master, origin/HEAD)
Author: Herbert Poul <herbert@codeux.design>
Date:   Mon Oct 19 15:04:17 2020 +0200

3. installation update
AuthPass-setup-1.7.9_1605.exe

**Results**
- It seems that installed application stars for the first time (no doubble run needed anymore).
- Dependencies errors in VSCode disappeared.
- Development.dart starts in VSCode for the first time.
- Running development.dart does not complain about the wrong kdbx password inspite of it is correct for sure.
- Main.dart does not run correctly (blank white screen) in VSCode (please see the attachment).
![2020-10-19 18_31_52-main dart - authpass - Visual Studio Code - Insiders](https://user-images.githubusercontent.com/43816134/96482335-0f5ca680-123c-11eb-979d-d07bc1b64168.png)

**Comment by hpoul** [comment]
@fialky as the error tries to indicate, you have to run with `lib/env/development.dart` (or any `lib/env/*.dart`), not `lib/main.dart` .. ie you must use the `-t` flag for `flutter run/build`.

**Comment by fialky** [comment]
Ah, OK. So, which exact command should be used for a production build (not debug build)?

**Comment by hpoul** [comment]
@fialky `flutter build windows -t lib/env/development.dart` can be used for a "release" build.. the difference is just API secrets (google drive, dropbox, one drive) and e.g. google analytics tracking id.. if you care about those you would have to sign up for api keys and configure them accordingly by creating your own `Env` subclass

**Comment by fialky** [comment]
Thx, release build made with your command runs well. (The main difference is in produced data/app.so file. Your command `flutter build windows -t lib/env/development.dart` made 12,9 MB app.so file. If I replace previous 733 KB app.so [generated by `flutter run --release`] file with the bigger one Authpass runs well without blank white screen anymore.)

**Comment by hpoul** [comment]
@fialky perfect  👍️ , thanks for the thorough testing. Yeah, I think this should solve this problem :-) (i think the current master should be basically the same as the released `1.7.9` from a few days ago.. so let's hope that this makes AuthPass usable on windows again ;-)
